app_dart/lib/src/model/firestore/ci_staging.dart

```diff
@@ -0,0 +1,157 @@
+  }) async {
+    final Document document = await firestoreService.getDocument(documentName);
+    return CiStaging.fromDocument(ciStagingDocument: document);
+  }
+
+  /// Create [CiStaging] from a Commit Document.
+  static CiStaging fromDocument({
+    required Document ciStagingDocument,
+  }) {
+    return CiStaging()
+      ..fields = ciStagingDocument.fields!
+      ..name = ciStagingDocument.name!;
+  }
+
+  /// The remaining number of checks in this staging.
+  int? get remaining => int.tryParse(fields![kRemainingField]!.integerValue ?? '');
+
+  /// The total number of checks in this staging.
+  int? get total => int.tryParse(fields![kTotalField]!.integerValue ?? '');
+
+  /// Mark a [checkRun] for a given [stage] with [conclusion].
+  ///
+  /// Returns a [StagingConclusion] record or throws. If the check_run was
+  /// both valid and recorded successfully, the record's `remaining` value
+  /// signals how many more tests are running. Returns the record (valid: false)
+  /// otherwise.
+  static Future<StagingConclusion> markConclusion({
+    required FirestoreService firestoreService,
+    // required String documentNameString,
+    required RepositorySlug slug,
+    required String sha,
+    required String stage,
+    required String checkRun,
+    required String conclusion,
+  }) async {
+    final logCrumb = 'markConclusion(${slug.owner}_${slug.name}_${sha}_$stage, $checkRun, $conclusion)';
+
+    // Marking needs to happen while in a transaction to ensure `remaining` is
+    // updated correctly. For that to happen correctly; we need to as perform a
+    // read of the document in the transaction as well. So start the transaction
+    // first thing.
+    final docRes = await firestoreService.documentResource();
+    final transactionResponse = await docRes.beginTransaction(
+      BeginTransactionRequest(
+        options: TransactionOptions(
+          readWrite: ReadWrite(),
+        ),
+      ),
+      kDatabase,
+    );
+    final transaction = transactionResponse.transaction;
+    if (transaction == null) {
+      log.warning('$logCrumb: transaction was null');
+      throw '$logCrumb: transaction was null';
+    }
+
+    var newRemaining = -1;
+
+    late final Document doc;
+
+    // transaction block
+    try {
+      // First: read the fields we want to change.
+      final documentPath = documentNameFor(sha: sha, slug: slug, stage: stage);
+      doc = await docRes.get(documentPath, mask_fieldPaths: [kRemainingField, checkRun], transaction: transaction);
+
+      // Fields and remaining _must_ be present.
+      if (doc.fields == null ||
+          doc.fields![kRemainingField] == null ||
+          doc.fields![kRemainingField]!.integerValue == null) {
+        log.warning('$logCrumb: missing field for $transaction / ${doc.fields}');
+        throw '$logCrumb missing fields';
+      }
+
+      final remaining = int.parse(doc.fields![kRemainingField]!.integerValue!);
+      newRemaining = remaining - 1;
```

Comment:
Should we have protections against this going negative?

Reply:
Added a log / throw above. This would be a bad state.  The calling function should only look for valid == true and remaining == 0.